src/libcore/iter/range.rs

```diff
@@ -5,169 +5,291 @@ use crate::usize;
 
 use super::{FusedIterator, TrustedLen};
 
-/// Objects that can be stepped over in both directions.
+/// Objects that have a notion of *successor* and *predecessor*
+/// for the purpose of range iterators.
 ///
-/// The `steps_between` function provides a way to efficiently compare
-/// two `Step` objects.
+/// This trait is `unsafe` because implementations of the `unsafe`
```

Comment:
I think this should be in a `# Safety` section.

Also, if this is an `unsafe trait` it feels like it should also have requirements on `Ord`, like `a.successor() > a` (if the successor exists).

Reply:
We have

```text
For any `a`, `n`:
* a.forward(n).unwrap() >= a
* a.backward(n).unwrap() <= a
```

This does allow for an exotic implementation where the **successor** operation is non-advancing, but I don't see a specific reason to rule that out.

But I definitely agree that it should be required for **forward** to be towards "larger" values and **backward** to be towards "smaller" values as defined by `Ord`. `Range` actually already relies on that for its correctness.

Reply:
One other thing that talking about `PartialOrd` makes me wonder: do we want to leave open the design space such that we could have `(0.0_f32).predecessor()` be `-0.0_f32`?

Reply:
Currently it is, as the requirement is only `a.forward(n) >= a` and the corollary `a.successor() >= a`. There's currently no requirement that `a.successor() != a`. (Equivalently, the other direction.)

It might be interesting to allow stepping floats through all representable floats (some routines really do want to test all concrete f32 values, after all), but I don't know whether it should be done for the unwrapped float types.